### PR TITLE
Pattern variable is not recognized in AND_AND_Expression

### LINKED ISSUE #1759: Pattern variable is not recognized in AND_AND_Expression
The related issue - https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/2992
Steps to reproduce:

- create the following class
```
import java.util.ArrayList;
import java.util.List;
public class X {
    public void test() {
        Object obj = "test"; 
        List<String> values = new ArrayList<>();
        if (obj instanceof String str && values.stream().anyMatch(str::equalsIgnoreCase)) {
            System.out.println(str);
        }
    }
}
```
- JDT returns
```
The method anyMatch(java.util.function.Predicate<? super java.lang.String>) in the type Stream<String> is not applicable for the arguments (java.util.function.Predicate<? super java.lang.String>)
```
The following code works fine
```
import java.util.ArrayList;
import java.util.List;
public class X {
    public void test() {
        Object obj = "test"; 
        List<String> values = new ArrayList<>();
        if (obj instanceof String str) {
        	if (values.stream().anyMatch(str::equalsIgnoreCase)) {
        		System.out.println(str);
        	}
        }
    }
}
```

### PR DESCRIPTION
**Pull request by snjeza**
Fixes #1759 

str is not recongized in
 `if (obj instanceof String str && values.stream().anyMatch(str::equalsIgnoreCase))`

### CODE REVIEW

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReferenceExpression.java [hunk 1]
```diff
@@ -998,7 +998,20 @@ private ReferenceExpression cachedResolvedCopy(TypeBinding targetType) {
...
+			// https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1759
+			LocalVariableBinding[] locals = this.enclosingScope.locals;
+			if (locals != null) {
+				LocalVariableBinding[] copyLocals = new LocalVariableBinding[locals.length];
+				System.arraycopy(locals, 0, copyLocals, 0, locals.length);
+				this.enclosingScope.locals = copyLocals;
+				for (LocalVariableBinding local: copyLocals) {
+					if (local != null) {
+						local.modifiers = local.modifiers & (1 << ExtraCompilerModifiers.AccPatternVariable);
+					}
```
Comment (srikanth-sankaran) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReferenceExpression.java:1010]: @snjeza This seems to be a side effect of unknown nature. What exactly is attempted here ?? Why would you want to adjust the modifiers thus ??

**Comment by srikanth-sankaran** [comment]
I think the real problem is `org.eclipse.jdt.internal.compiler.ast.ReferenceExpression.copy()` not propagating the pattern variables when true/false to the copy from the original

**Comment by srikanth-sankaran** [comment]
I have created an alternate PR here: https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1768 - let us see what the test results look like

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReferenceExpression.java (hunk 1 above)
Comment (snjeza) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReferenceExpression.java:1010]: I agree. #1768 is better.

**Comment by srikanth-sankaran** [comment]
Closed in favor of https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1768

**Pull request closed without merging**